packages/flutter_tools/lib/src/fuchsia/fuchsia_pm.dart

```diff
@@ -109,26 +77,168 @@ class FuchsiaPM {
+  /// the 'serve' command.
+  Future<bool> newrepo(String repoPath) {
+    return _runPMCommand(<String>[
+      'newrepo',
+      '-repo',
+      repoPath,
+    ]);
+  }
+
+  /// Spawns an http server in a new process for serving Fuchisa packages.
+  ///
+  /// The arguemnt [repoPath] should have previously been an arguemnt to
+  /// [newrepo]. The [host] should be the host reported by
+  /// [FuchsiaDevFinder.resolve], and [port] should be an unused port for the
+  /// http server to bind.
+  Future<Process> serve(String repoPath, String host, int port) async {
+    if (fuchsiaArtifacts.pm == null) {
+      throwToolExit('Fuchsia pm tool not found');
+    }
+    final List<String> command = <String>[
+      fuchsiaArtifacts.pm.path,
+      'serve',
+      '-repo',
+      repoPath,
+      '-l',
+      '$host:$port',
     ];
-    printTrace("Running: '${command.join(' ')}'");
-    final ProcessResult result = await processManager.run(command);
-    if (result.exitCode != 0) {
-      printError('Error archiving Fuchsia package: ');
-      printError(result.stdout);
-      printError(result.stderr);
+    final Process process = await runCommand(command);
+    process.stdout
+        .transform(utf8.decoder)
+        .transform(const LineSplitter())
+        .listen(printTrace);
+    process.stderr
+        .transform(utf8.decoder)
+        .transform(const LineSplitter())
+        .listen(printError);
+    return process;
+  }
+
+  /// Publishes a Fuchsia package to a served package repository.
+  ///
+  /// For a package repo initialized with [newrepo] at [repoPath] and served
+  /// by [serve], this call publishes the `far` package at [packagePath] to
+  /// the repo such that it will be visible to devices connecting to the
+  /// package server.
+  Future<bool> publish(String repoPath, String packagePath) {
+    return _runPMCommand(<String>[
+      'publish',
+      '-a',
+      '-r',
+      repoPath,
+      '-f',
+      packagePath,
+    ]);
+  }
+
+  Future<bool> _runPMCommand(List<String> args) async {
+    if (fuchsiaArtifacts.pm == null) {
+      throwToolExit('Fuchsia pm tool not found');
+    }
+    final List<String> command = <String>[fuchsiaArtifacts.pm.path] + args;
+    final RunResult result = await runAsync(command);
+    return result.exitCode == 0;
+  }
+}
+
+/// A class for running and retaining state for a Fuchsia package server.
+///
+/// [FuchsiaPackageServer] takes care of initializing the package repository,
+/// spinning up the package server, publishing packages, and shutting down the
+/// the server.
+///
+/// Example usage:
+/// var server = FuchsiaPackageServer(
+///     '/path/to/repo',
+///     await FuchsiaDevFinder.resolve(deviceName),
+///     await freshPort());
+/// try {
+///   await server.start();
+///   await server.addPackage(farArchivePath);
+///   ...
+/// } finally {
+///   server.stop();
+/// }
+class FuchsiaPackageServer {
+  FuchsiaPackageServer(
+    String repo,
```

Comment:
For positional parameters I would use `this._repo` and such

Reply:
Done.